### PR TITLE
Togroups policy fixup

### LINKED ISSUE #13083: #12920 toGroups policy fix follow ups
- [ ] Refactor `createDerivative{CNP,CCNP}` functions - https://github.com/cilium/cilium/pull/12920#discussion_r483008050
- [ ] Replace for loop logic with controller in `addDerivativeCNP` - https://github.com/cilium/cilium/pull/12920#discussion_r483008873
- [ ] Pass proper contexts from controllers to functions in `pkg/policy/groups` - https://github.com/cilium/cilium/pull/12920#discussion_r482078653

### PR DESCRIPTION
**Pull request by psinghal20**
This change resolves the comments for followup fixes associated with PR https://github.com/cilium/cilium/pull/12920
- Refactors out API rules modification from createDerivative{CNP, CCNP} functions
- Pass proper contexts from the caller controller functions

I am not sure of the 2nd task mentioned on the issue regarding `Replacing for loop with controller for increasing reliability`, the `addDerivativePolicy` function already runs as a controller in caller functions. Also, the for loop returns CCNP and CNP objects which are used later. @aanm can you please take a look and clarify what you meant [here](https://github.com/cilium/cilium/pull/12920#discussion_r483008873)?

Fixes: https://github.com/cilium/cilium/issues/13083

### CODE REVIEW

**Comment by aanm** [comment]
@psinghal20 thanks for taking a look at this. You are correct, the caller of this function is already a controller. Then I would suggest to remove the for loop `for numAttempts := 0; numAttempts <= maxNumberOfAttempts; numAttempts++ {` as it is redundant.

**Comment by psinghal20** [comment]
Hi @aanm, can you please confirm if returning the `derivativeErr` to retry would be fine here? I read the comments regarding this for loop [here](https://github.com/cilium/cilium/blob/master/pkg/policy/groups/actions.go#L228-L232), and I am not sure if we should return an error here and let the controller retry continuously even with exponential backoff.

**Comment by aanm** [comment]
> Hi @aanm, can you please confirm if returning the `derivativeErr` to retry would be fine here? I read the comments regarding this for loop [here](https://github.com/cilium/cilium/blob/master/pkg/policy/groups/actions.go#L228-L232), and I am not sure if we should return an error here and let the controller retry continuously even with exponential backoff.

Looking at the code it makes more sense to return `derivativeErr` if, after the for loop, it is not nil. Otherwise there won't be any retry to recreate the derivative CNP.

**Comment by psinghal20** [comment]
Hey @aanm, I have removed the for loop retry logic and return `derivativeErr` for controller retry logic. Please take a look. Thanks!

File: pkg/policy/groups/actions.go [hunk 1]
```diff
@@ -225,33 +225,25 @@ func addDerivativePolicy(ctx context.Context, cnp *cilium_v2.CiliumNetworkPolicy
...
 	// error.
-	for numAttempts := 0; numAttempts <= maxNumberOfAttempts; numAttempts++ {
-		if clusterScoped {
-			derivativeCCNP, derivativeErr = createDerivativeCCNP(ctx, cnp)
-			derivativePolicy = derivativeCCNP
-		} else {
-			derivativeCNP, derivativeErr = createDerivativeCNP(ctx, cnp)
-			derivativePolicy = derivativeCNP
-		}
+	if clusterScoped {
```
Comment (aanm) [inline pkg/policy/groups/actions.go:223]: The `maxNumberOfAttempts` can be removed from the top of this file as well.

Comment (psinghal20) [reply pkg/policy/groups/actions.go:223]: Done.

**Comment by jrajahalme** [comment]
Checkpatch was not happy about the commit subject lines:
```
=========================================================
[0/3] Running on 657ee1d6f5a3c896e5242491aa0e9e7178fcf7ab
policy/groups: refactor out api rules modification logic in createDerivative{CNP,CCNP} to a separate dedicated function
=========================================================
ERROR:CUSTOM: Please avoid long commit subjects (max: 75, found: 119)
=========================================================
[1/3] Running on 164344874e36a66fe76b3efd2e92d2be7c8146f7
policy/groups: pass proper context from caller function to deleteDerivative{CNP, CCNP}
=========================================================
ERROR:CUSTOM: Please avoid long commit subjects (max: 75, found: 86)
=========================================================
[2/3] Running on d0b83e2398c51777e145ad94f9f621bcd8c2b73f
policy/groups: remove retry login from addDerivativePolicy method in favor of controller retry logic
=========================================================
ERROR:CUSTOM: Please avoid long commit subjects (max: 75, found: 100)
```

**Comment by psinghal20** [comment]
Hi @jrajahalme, I have reworded the commits. Hopefully, the check should pass now. :slightly_smiling_face:

**Comment by jrajahalme** [comment]
test-me-please

**Comment by jrajahalme** [comment]
test-1.20-4.19

**Comment by jrajahalme** [comment]
Known test-1.21-4.9 flakes #15097 & #16127

**Comment by jrajahalme** [comment]
test-1.21-4.9

**Comment by jrajahalme** [comment]
All the test flakes are known flakes unrelated to the changes in this PR. Marking as ready-to-merge.

**Review by jrajahalme — APPROVED** [review]

**Pull request merged**